Fauna/Core/Connection.cs

```diff
@@ -57,44 +60,63 @@ public async IAsyncEnumerable<Event<T>> OpenStream<T>(
+                {
+                    var streamData = new MemoryStream();
+                    stream.Serialize(streamData);
 
-                await using var streamAsync = await response.Content.ReadAsStreamAsync(cancellationToken);
-                using var streamReader = new StreamReader(streamAsync);
+                    var response = await _cfg.HttpClient
+                        .SendAsync(
+                            CreateHttpRequest(path, streamData, headers),
+                            HttpCompletionOption.ResponseHeadersRead,
+                            cancellationToken)
+                        .ConfigureAwait(false);
 
-                while (!streamReader.EndOfStream && !cancellationToken.IsCancellationRequested)
-                {
-                    string? line = await streamReader.ReadLineAsync().WaitAsync(cancellationToken);
-                    if (string.IsNullOrWhiteSpace(line))
+                    if (!response.IsSuccessStatusCode)
                     {
-                        continue;
+                        // TRICKY: we need to break the events listener loop
+                        listener.Dispatch(null);
+                        return response;
                     }
 
-                    var evt = Event<T>.From(line, ctx);
-                    stream.LastCursor = evt.Cursor;
-                    listener.Dispatch(evt);
-                }
+                    await using var streamAsync = await response.Content.ReadAsStreamAsync(cancellationToken);
+                    using var streamReader = new StreamReader(streamAsync);
 
-                listener.Close();
-                return response;
-            });
+                    while (!streamReader.EndOfStream && !cancellationToken.IsCancellationRequested)
+                    {
+                        string? line = await streamReader.ReadLineAsync().WaitAsync(cancellationToken);
+                        if (string.IsNullOrWhiteSpace(line))
+                        {
+                            continue;
+                        }
+
+                        var evt = Event<T>.From(line, ctx);
+                        stream.LastCursor = evt.Cursor;
+                        listener.Dispatch(evt);
+                    }
+
+                    listener.Close();
+                    return response;
+                });
 
             await foreach (var evt in listener.Events().WithCancellation(cancellationToken))
             {
+                if (evt is null) break;
+
                 yield return evt;
             }
 
             await streamTask;
+            if (streamTask.Result.Result.IsSuccessStatusCode)
+            {
+                continue;
+            }
+
+            var httpResponse = streamTask.Result.Result;
+            string body = await httpResponse.Content.ReadAsStringAsync(cancellationToken);
+
+            throw ExceptionFactory.FromRawResponse(body, httpResponse);
```

Comment:
This ordering kinda threw me for a sec; maybe handle the `!IsSuccessStatusCode` and let the `continue` be implied by falling through?
```suggestion
            if (!streamTask.Result.Result.IsSuccessStatusCode)
            {
                var httpResponse = streamTask.Result.Result;
                string body = await httpResponse.Content.ReadAsStringAsync(cancellationToken);

                throw ExceptionFactory.FromRawResponse(body, httpResponse);
            }
```

Reply:
![image](https://github.com/user-attachments/assets/c8eb1b42-116f-4ace-8e65-6b6c3b0a4048)
I agree with Rider on this one, I find early returns more readable

Reply:
🤯 
Rider betrayed me.  IntelliJ?!  More like DumbJ.